app/src/behaviors/behavior_combo.c

```diff
@@ -0,0 +1,471 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_combo
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <sys/dlist.h>
+#include <kernel.h>
+
+#include <zmk/behavior.h>
+#include <zmk/matrix.h>
+#include <zmk/endpoints.h>
+#include <zmk/event-manager.h>
+#include <zmk/events/position-state-changed.h>
+#include <zmk/events/keycode-state-changed.h>
+#include <zmk/events/modifiers-state-changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+struct combo_t {
+    int32_t key_positions[CONFIG_ZMK_BHV_COMBO_MAX_KEYS_PER_COMBO];
+    int32_t key_position_len;
+    struct zmk_behavior_binding behavior;
+    int32_t timeout_ms;
+    // if slow release is set, the combo releases when the last key is released.
+    // otherwise, the combo releases when the first key is released.
+    bool slow_release;
+    // the virtual key position is a key position outside the range used by the keyboard.
+    // it is necessary so hold-taps can uniquely identify a behavior.
+    int32_t virtual_key_position;
+};
+
+struct active_combo {
+    struct combo_t *combo;
+    // key_positions_pressed is filled with key_positions when the combo is pressed.
+    // The keys are removed from this array when they are released.
+    // Once this array is empty, the behavior is released.
+    struct position_state_changed *key_positions_pressed[CONFIG_ZMK_BHV_COMBO_MAX_KEYS_PER_COMBO];
+};
+
+struct combo_candidate {
+    struct combo_t *combo;
+    // the time after which this behavior should be removed from candidates.
+    // by keeping track of when the candidate should be cleared there is no
+    // possibility of accidental releases.
+    int64_t timeout_at;
+};
+
+// set of keys pressed
+struct position_state_changed *pressed_keys[CONFIG_ZMK_BHV_COMBO_MAX_KEYS_PER_COMBO] = {NULL};
+// the set of candidate combos based on the currently pressed_keys
+struct combo_candidate candidates[CONFIG_ZMK_BHV_COMBO_MAX_COMBOS_PER_KEY];
+// the last candidate that was completely pressed
+struct combo_t *fully_pressed_combo = NULL;
+// a lookup dict that maps a key position to all combos on that position
+struct combo_t *combo_lookup[ZMK_KEYMAP_LEN][CONFIG_ZMK_BHV_COMBO_MAX_COMBOS_PER_KEY] = {NULL};
+// combos that have been activated and still have (some) keys pressed
+// this array is always contiguous from 0.
+struct active_combo active_combos[CONFIG_ZMK_BHV_COMBO_MAX_PRESSED_COMBOS] = {NULL};
+int active_combo_count = 0;
+
+struct k_delayed_work timeout_task;
+int64_t timeout_task_timeout_at;
+
+struct behavior_combo_data {};
+static struct behavior_combo_data behavior_combo_data;
```

Comment:
You can ditch this altogether if not using it.

Reply:
Yeah, I was just thinking about this today. The combos arent behaviors at all, they shouldn't even be in the behaviors dir. They should be on the same level as a keymap.